cmd/docker-app/install.go

```diff
@@ -0,0 +1,333 @@
+	credentialsets   []string
+}
+
+type parametersOptions struct {
+	parametersFiles []string
+	env             []string
+}
+
+func (o *parametersOptions) addFlags(flags *pflag.FlagSet) {
+	flags.StringArrayVarP(&o.parametersFiles, "parameters-files", "f", []string{}, "Override parameters files")
+	flags.StringArrayVarP(&o.env, "set", "s", []string{}, "Override parameters values")
+}
+
+type nameKind uint
+
+const (
+	_ nameKind = iota
+	nameKindEmpty
+	nameKindFile
+	nameKindDir
+	nameKindReference
+)
+
+const longDescription = `Install the application on either Swarm or Kubernetes.
+Bundle name is optional, and can be
+- empty: resolve to any *.dockerapp file or directory in working dir
+- existing file path: work with any *.dockerapp file or dir, or any CNAB bundle file (signed or unsigned)
+- match a bundle name in the local duffle bundle repository
+- refers to a CNAB bundle in a container registry
+`
+
+// installCmd represents the install command
+func installCmd(dockerCli command.Cli) *cobra.Command {
+	var opts installOptions
+
+	cmd := &cobra.Command{
+		Use:     "install [<bundle name>] [options]",
+		Aliases: []string{"deploy"},
+		Short:   "Install an application",
+		Long:    longDescription,
+		Args:    cli.RequiresMaxArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runInstall(dockerCli, firstOrEmpty(args), opts)
+		},
+	}
+	opts.parametersOptions.addFlags(cmd.Flags())
+	cmd.Flags().StringVarP(&opts.orchestrator, "orchestrator", "o", "", "Orchestrator to install on (swarm, kubernetes)")
+	cmd.Flags().StringVar(&opts.namespace, "namespace", "", "Namespace to use (default: namespace in metadata)")
+	cmd.Flags().StringVar(&opts.kubeNamespace, "kubernetes-namespace", "default", "Kubernetes namespace to install into")
+	cmd.Flags().StringVar(&opts.stackName, "name", "", "Installation name (defaults to application name)")
+	cmd.Flags().StringVar(&opts.targetContext, "target-context", "", "Context on which to install the application")
+	cmd.Flags().BoolVar(&opts.insecure, "insecure", false, "Use insecure registry, without SSL")
+	cmd.Flags().BoolVar(&opts.sendRegistryAuth, "with-registry-auth", false, "Sends registry auth")
+	cmd.Flags().StringArrayVarP(&opts.credentialsets, "credential-set", "c", []string{}, "Use a duffle credentialset (either a YAML file, or a credential set present in the duffle credential store)")
+
+	return cmd
+}
+
+func runInstall(dockerCli command.Cli, appname string, opts installOptions) error {
+	muteDockerCli(dockerCli)
+	targetContext := getTargetContext(opts.targetContext)
+	parameterValues, err := prepareParameters(opts.parametersOptions)
+	if err != nil {
+		return err
+	}
+
+	bndl, err := resolveBundle(dockerCli, opts.namespace, appname)
+	if err != nil {
+		return err
+	}
+	if opts.sendRegistryAuth {
+		return errors.New("with-registry-auth is not supported at the moment")
+	}
+	if err := bndl.Validate(); err != nil {
+		return err
+	}
+	h := duffleHome()
```

Comment:
I think there is room for extracting some code here in testable function :angel: